File: cwclock-mobile/lib/screens/organizations/members_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';

import '../../api/api_client.dart';
import '../../i18n/app_localizations.dart';
import '../../models/member.dart';
import '../../providers/locale_provider.dart';
import '../../providers/organizations_provider.dart';
import '../../providers/permissions.dart' as perm;
import '../../providers/session_provider.dart';
import '../../theme.dart';
import '../../widgets/app_button.dart';

const _roles = ['admin', 'member', 'reader'];

/// List members with their role; invite (owner-only), change an existing
/// member's role (owner-only), remove a member (admin/owner). The owner's
/// own row exposes neither action. Ported from
/// src/screens/organizations/MembersScreen.js.
class MembersScreen extends ConsumerStatefulWidget {
  const MembersScreen({super.key});

  @override
  ConsumerState<MembersScreen> createState() => _MembersScreenState();
}

class _MembersScreenState extends ConsumerState<MembersScreen> {
  @override
  void initState() {
    super.initState();
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId != null) {
      Future.microtask(() => ref.read(organizationsProvider.notifier).listMembers(orgId));
    }
  }

  String _roleLabel(String role, String Function(String, [Map<String, String>?]) t) {
    final capitalized = '${role[0].toUpperCase()}${role.substring(1)}';
    return t('organizations.role$capitalized');
  }

  void _handleChangeRole(Member member) {
    final locale = ref.read(localeProvider);
    final t = translateWith(locale);
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null) return;

    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text(t('organizations.changeRole')),
        content: Text(member.email),
        actions: [
          for (final role in _roles)
            TextButton(
              onPressed: () async {
                Navigator.pop(context);
                try {
                  await ref.read(organizationsProvider.notifier).updateMemberRole(orgId, member.userId, role);
                } catch (e) {
                  if (mounted) {
                    _showError(apiErrorMessage(asApiException(e), locale));
                  }
                }
              },
              child: Text(_roleLabel(role, t)),
            ),
          TextButton(onPressed: () => Navigator.pop(context), child: Text(t('common.cancel'))),
        ],
      ),
    );
  }

  void _showError(String message) {
    final locale = ref.read(localeProvider);
    final t = translateWith(locale);
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text(t('common.retry')),
        content: Text(message),
        actions: [TextButton(onPressed: () => Navigator.pop(context), child: Text(t('common.close')))],
      ),
    );
  }

  void _handleRemove(Member member) {
    final locale = ref.read(localeProvider);
    final t = translateWith(locale);
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null) return;

    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text(t('organizations.removeMemberTitle')),
        content: Text(t('organizations.removeMemberBody', {'name': member.email})),
        actions: [
          TextButton(onPressed: () => Navigator.pop(context), child: Text(t('common.cancel'))),
          TextButton(
            onPressed: () async {
              Navigator.pop(context);
              try {
                await ref.read(organizationsProvider.notifier).removeMember(orgId, member.userId);
              } catch (e) {
                if (mounted) _showError(apiErrorMessage(asApiException(e), locale));
              }
            },
            child: Text(t('organizations.removeMember'), style: TextStyle(color: AppColors.of(context).danger)),
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final locale = ref.watch(localeProvider);
    final t = translateWith(locale);
    final session = ref.watch(sessionProvider);
    final orgsState = ref.watch(organizationsProvider);

    final org = orgsState.items.where((o) => o.id == session.orgId).firstOrNull;
    final canInvite = perm.isOrgOwner(session.user, org);
    final canRemove = perm.isAdminOrOwner(session.user, orgsState.members);
    final members = orgsState.members;

    return Scaffold(
      appBar: AppBar(title: Text(t('organizations.membersTitle'))),
      body: SafeArea(
        child: Column(
          children: [
            if (canInvite)
              Padding(
                padding: EdgeInsets.all(AppSpacing.of(2)),
                child: AppButton(
                  title: t('organizations.inviteMember'),
                  onPressed: () => context.push('/members/invite'),
                ),
              ),
            Expanded(
              child: members.isEmpty
                  ? Padding(
                      padding: EdgeInsets.symmetric(horizontal: AppSpacing.of(2)),
                      child: Text(t('organizations.noMembers'), style: TextStyle(color: AppColors.of(context).textMuted)),
                    )
                  : ListView.separated(
                      padding: EdgeInsets.symmetric(horizontal: AppSpacing.of(2)),
                      itemCount: members.length,
                      separatorBuilder: (_, _) => Divider(height: 1, color: AppColors.of(context).border),
                      itemBuilder: (context, index) {
                        final member = members[index];
                        final isOwnerRow = member.role == 'owner';
                        final displayName = (member.name?.isNotEmpty ?? false) || (member.surname?.isNotEmpty ?? false)
                            ? '${member.name ?? ''} ${member.surname ?? ''}'.trim()
                            : member.email;
                        return ListTile(
                          contentPadding: EdgeInsets.zero,
                          title: Text(displayName, maxLines: 1, overflow: TextOverflow.ellipsis),
                          subtitle: Text(
                            '${member.email} · ${_roleLabel(member.role, t)}',
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                          ),
                          trailing: isOwnerRow
                              ? null
                              : Row(
                                  mainAxisSize: MainAxisSize.min,
                                  children: [
                                    if (canInvite)
                                      TextButton(
                                        onPressed: () => _handleChangeRole(member),
                                        child: Text(t('organizations.changeRole')),
                                      ),
                                    if (canRemove)
                                      TextButton(
                                        onPressed: () => _handleRemove(member),
                                        child: Text(
                                          t('organizations.removeMember'),
                                          style: TextStyle(color: AppColors.of(context).danger),
                                        ),
                                      ),
                                  ],
                                ),
                        );
                      },
                    ),
            ),
          ],
        ),
      ),
    );
  }
}

extension _FirstOrNull<T> on Iterable<T> {
  T? get firstOrNull {
    final it = iterator;
    return it.moveNext() ? it.current : null;
  }
}

File: cwclock-mobile/src/screens/organizations/MembersScreen.js
import React, { useEffect } from "react";
import { Alert, FlatList, StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import Screen from "../../components/Screen";
import Button from "../../components/Button";
import { useI18n } from "../../i18n/I18nContext";
import { apiErrorMessage } from "../../i18n/translate";
import theme from "../../theme";
import { isOrgOwner, isAdminOrOwner } from "../../common/permissions";
import { listMembersApi, updateMemberRoleApi, removeMemberApi } from "../../redux/organizations/organizations.actions";

const ROLES = ["admin", "member", "reader"];

// MembersScreen: the "user RBAC in the organization" feature from
// ai-instruct-106 - list members with their role, invite (owner-only,
// InviteMemberScreen), change an existing member's role (owner-only - the
// API supports this via UpdateMember, but cwclock-ui's web app has no UI for
// it, only invite-time role, see organizations.actions.js's
// updateMemberRoleApi comment) and remove a member (admin/owner). The
// owner's own row exposes neither action - removing/demoting the owner
// isn't a normal RBAC operation.
const MembersScreen = ({ navigation }) => {
  const { t, locale } = useI18n();
  const dispatch = useDispatch();
  const { orgId, user } = useSelector((state) => state.session);
  const { items: orgs, members } = useSelector((state) => state.organizations);

  const org = orgs.find((o) => o.id === orgId);
  const canInvite = isOrgOwner(user, org);
  const canRemove = isAdminOrOwner(user, members);

  useEffect(() => {
    if (orgId) dispatch(listMembersApi(orgId));
  }, [orgId, dispatch]);

  const roleLabel = (role) => t(`organizations.role${role.charAt(0).toUpperCase()}${role.slice(1)}`);

  const handleChangeRole = (member) => {
    Alert.alert(
      t("organizations.changeRole"),
      member.email,
      ROLES.map((role) => ({
        text: roleLabel(role),
        onPress: () => dispatch(updateMemberRoleApi(orgId, member.userId, role)).catch((e) => Alert.alert(t("common.retry"), apiErrorMessage(e, locale))),
      })).concat([{ text: t("common.cancel"), style: "cancel" }])
    );
  };

  const handleRemove = (member) => {
    Alert.alert(t("organizations.removeMemberTitle"), t("organizations.removeMemberBody", { name: member.email }), [
      { text: t("common.cancel"), style: "cancel" },
      {
        text: t("organizations.removeMember"),
        style: "destructive",
        onPress: () =>
          dispatch(removeMemberApi(orgId, member.userId)).catch((e) =>
            Alert.alert(t("common.retry"), apiErrorMessage(e, locale))
          ),
      },
    ]);
  };

  return (
    <Screen scroll={false}>
      {canInvite && (
        <View style={styles.header}>
          <Button title={t("organizations.inviteMember")} onPress={() => navigation.navigate("InviteMember")} />
        </View>
      )}
      <FlatList
        data={members}
        keyExtractor={(m) => m.userId}
        contentContainerStyle={styles.list}
        ListEmptyComponent={<Text style={styles.empty}>{t("organizations.noMembers")}</Text>}
        renderItem={({ item }) => {
          const isOwnerRow = item.role === "owner";
          return (
            <View style={styles.row}>
              <View style={styles.rowMain}>
                <Text style={styles.rowText} numberOfLines={1}>
                  {item.name || item.surname ? `${item.name} ${item.surname}`.trim() : item.email}
                </Text>
                <Text style={styles.rowSub} numberOfLines={1}>
                  {item.email} · {roleLabel(item.role)}
                </Text>
              </View>
              {!isOwnerRow && (
                <View style={styles.rowActions}>
                  {canInvite && (
                    <TouchableOpacity onPress={() => handleChangeRole(item)} style={styles.actionBtn}>
                      <Text style={styles.actionText}>{t("organizations.changeRole")}</Text>
                    </TouchableOpacity>
                  )}
                  {canRemove && (
                    <TouchableOpacity onPress={() => handleRemove(item)} style={styles.actionBtn}>
                      <Text style={styles.actionTextDanger}>{t("organizations.removeMember")}</Text>
                    </TouchableOpacity>
                  )}
                </View>
              )}
            </View>
          );
        }}
      />
    </Screen>
  );
};

const styles = StyleSheet.create({
  header: {
    padding: theme.spacing(2),
  },
  list: {
    paddingHorizontal: theme.spacing(2),
  },
  row: {
    paddingVertical: theme.spacing(1.5),
    borderBottomWidth: 1,
    borderBottomColor: theme.color.border,
  },
  rowMain: {
    marginBottom: theme.spacing(0.5),
  },
  rowText: {
    fontSize: 16,
    color: theme.color.text,
  },
  rowSub: {
    fontSize: 13,
    color: theme.color.textMuted,
    marginTop: 2,
  },
  rowActions: {
    flexDirection: "row",
    gap: theme.spacing(2),
  },
  actionBtn: {
    paddingVertical: theme.spacing(0.5),
  },
  actionText: {
    color: theme.color.primary,
    fontSize: 13,
    fontWeight: "600",
  },
  actionTextDanger: {
    color: theme.color.danger,
    fontSize: 13,
    fontWeight: "600",
  },
  empty: {
    color: theme.color.textMuted,
    paddingVertical: theme.spacing(2),
  },
});

export default MembersScreen;
